executor/runtime/benchmark/operator.go

```diff
@@ -181,3 +188,152 @@ func (o *opSink) Next(ctx *runtime.TaskContext, r *runtime.Record, _ int) ([]run
 }
 
 func (o *opSink) NextWantedInputIdx() int { return 0 }
+
+type opProducer struct {
+	tid       int32
+	pk        int32
+	schemaVer int32
+	dataCnt   int32
+
+	ddlFrequency int32
+	outputCnt    int
+}
+
+var ddlLock sync.Mutex
+
+func (o *opProducer) Close() error { return nil }
+
+func (o *opProducer) Prepare() error { return nil }
+
+func (o *opProducer) NextWantedInputIdx() int { return runtime.DontNeedData }
+
+func (o *opProducer) Next(ctx *runtime.TaskContext, _ *runtime.Record, _ int) ([]runtime.Chunk, bool, error) {
+	outputData := make([]runtime.Chunk, o.outputCnt)
+	binlogID := 0
+	for i := 0; i < 128; i++ {
+		if o.pk >= o.dataCnt {
+			return outputData, true, nil
+		}
+		o.pk++
+		if o.pk == 10000 {
+			log.L().Info("got 10000 data")
+		}
+		if o.pk%o.ddlFrequency == 0 {
+			o.schemaVer++
+			ddlLock.Lock()
+			for i := range outputData {
+				payload := &pb.Record{
+					Tp:        pb.Record_DDL,
+					Tid:       o.tid,
+					SchemaVer: o.schemaVer,
+				}
+				r := runtime.Record{
+					Tid:     o.tid,
+					Payload: payload,
+				}
+				outputData[i] = append(outputData[i], &r)
+			}
+			ddlLock.Unlock()
+		}
+		payload := &pb.Record{
+			Tp:        pb.Record_Data,
+			Tid:       o.tid,
+			SchemaVer: o.schemaVer,
+			Pk:        o.pk,
+		}
+		r := runtime.Record{
+			Tid:     o.tid,
+			Payload: payload,
+		}
+		outputData[binlogID] = append(outputData[binlogID], &r)
+		binlogID = (binlogID + 1) % o.outputCnt
+	}
+	return outputData, false, nil
+}
+
+type opBinlog struct {
+	binlogChan chan *runtime.Record
+	wal        []*runtime.Record
+	addr       string
+
+	server      mock.GrpcServer
+	cacheRecord *runtime.Record
+	ctx         *runtime.TaskContext
+}
+
+func (o *opBinlog) Close() error {
+	o.server.Stop()
+	return nil
+}
+
+func (o *opBinlog) Prepare() (err error) {
+	o.binlogChan = make(chan *runtime.Record, 1024)
+	if test.GlobalTestFlag {
+		o.server, err = mock.NewTestServer(o.addr, o)
+	} else {
+		lis, err := net.Listen("tcp", o.addr)
+		if err != nil {
+			return err
+		}
+		s := grpc.NewServer()
+		pb.RegisterTestServiceServer(s, o)
+		go func() {
+			err1 := s.Serve(lis)
+			if err1 != nil {
+				log.L().Logger.Error("start grpc server failed", zap.Error(err))
+				panic(err1)
+			}
+		}()
+	}
+	return err
+}
+
+func (o *opBinlog) NextWantedInputIdx() int {
+	if o.cacheRecord != nil {
+		return runtime.DontNeedData
+	}
+	return runtime.DontRequireIndex
+}
+
+func (o *opBinlog) Next(ctx *runtime.TaskContext, r *runtime.Record, _ int) ([]runtime.Chunk, bool, error) {
```

Comment:
What is the return `[]runtime.Chunk` used for? seems it is always `nil`

Reply:
Oh, is it caused by interface definition, not all operators return chunk in `Next`

Reply:
yep. the design of interface has to be compatible